src/libsyntax/parse/parser.rs

```diff
@@ -2870,7 +2870,46 @@ impl<'a> Parser<'a> {
                 let (span, e) = self.interpolated_or_expr_span(e)?;
                 (lo.to(span), ExprKind::Box(e))
             }
-            _ => return self.parse_dot_or_call_expr(Some(attrs))
+            token::Ident(..) if self.token.is_ident_named("not") => {
+                // `not` is just an ordinary identifier in Rust-the-language,
+                // but as `rustc`-the-compiler, we can issue clever diagnostics
+                // for confused users who really want to say `!`
+                let token_cannot_continue_expr = |t: &token::Token| match *t {
+                    // These tokens can't continue an expression after an ident
+                    token::Ident(..) | token::Literal(..) | token::Pound => true,
```

Comment:
Also, it should be easier now to add the `Nt` variants from https://github.com/rust-lang/rust/pull/48858#issuecomment-373946361

Reply:
Also, for `token::Ident` the correct condition is `t is Token(ident) && ident_can_begin_expr(ident)`